File: 152.maximum-product-subarray.cpp
/*
 * @lc app=leetcode id=152 lang=cpp
 *
 * [152] Maximum Product Subarray
 */
#include <bits/stdc++.h>
using namespace std;


// @lc code=start
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        return moddedKadane(nums);
    }

    int moddedKadane(vector<int>& nums) {
        int currMinProd = 1;
        int currMaxProd = 1;
        int globalMax = nums[0];
        bool contains_zero = false;
        for (int num: nums) {
            if (num == 0) {
                currMaxProd = 1;
                currMinProd = 1;
                contains_zero = true;
                continue;
            }
            int temp = currMaxProd * num;
            currMaxProd = max(currMaxProd*num, max(num*currMinProd, num));
            currMinProd = min(temp, min(num*currMinProd, num));
            globalMax = max(globalMax, currMaxProd);
        }
        if (contains_zero && globalMax<0) { globalMax = 0; }
        return globalMax;
    }
};
// @lc code=end

File: 152.maximum-product-subarray.py
#
# @lc app=leetcode id=152 lang=python3
#
# [152] Maximum Product Subarray
#
from typing import List

# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        currMax, currMin = 1, 1
        res = nums[0]
        for num in nums:
            tmp = currMax * num
            currMax = max(currMax*num, currMin*num, num)
            currMin = min(tmp, currMin*num, num)
            res = max(res, currMax)
        return res
# @lc code=end

def main():
    s = Solution()
    nums = [-2, 0, -1]
    s.maxProduct(nums)

if __name__ == '__main__':
    main()